Sources/Core/Sources/ECS/Systems/PlayerInputSystem.swift

```diff
@@ -136,16 +136,60 @@ public final class PlayerInputSystem: System {
       if event.key == .enter {
         if !message.isEmpty {
           try connection?.sendPacket(ChatMessageServerboundPacket(message))
+          guiState.playerMessageHistory.append(message)
+          guiState.currentMessageIndex = nil
         }
         guiState.messageInput = nil
         return true
       } else if event.key == .escape {
         guiState.messageInput = nil
+        guiState.currentMessageIndex = nil
         return true
       } else if event.key == .delete {
         if !message.isEmpty {
           guiState.messageInput?.removeLast()
         }
+      } else if event.key == .upArrow {
+        // If no message is selected, select the above message
+        if guiState.currentMessageIndex == nil {
+          guiState.stashedMessageInput = guiState.messageInput
+          // If there are no messages, do nothing
+          if guiState.playerMessageHistory.count > 0 {
+            // Index up a message
+            guiState.currentMessageIndex = guiState.playerMessageHistory.count - 1
+            if let currentMessageIndex: Int = guiState.currentMessageIndex {
+              // Set the message input to the message at the index
+              guiState.messageInput = guiState.playerMessageHistory[guiState.currentMessageIndex!]
```

Comment:
You don't need to force unwrap `guiState.currentMessageIndex`, you can replace it with `currentMessageIndex`. Alternatively (and probably better), you can actually just store `count - 1` in an `index` variable at line 159 and then assign `index` to `guiState.currentMessageIndex` afterwards. That way you don't have to do any unwrapping (you can completely avoid unwrapping in this case because you're setting the value immediately beforehand, and you know it's not `nil`, the trick is just knowing how to convince the compiler).